README.md
<a href="https://www.hypertrack.com/">
    <img src="https://hypertrack-blog-img.s3-us-west-2.amazonaws.com/Green0svg.svg" alt="HyperTrack logo" title="HyperTrack" align="right" height="40" />
</a>

# Visits app (Android)

<p align="center">👉 <a href="https://github.com/hypertrack/visits-ios">Looking for the iOS version?</a></p>

This is an open-sourced logistics application (also available in [Google Play](https://play.google.com/store/apps/details?id=com.hypertrack.logistics.android.github)) built using HyperTrack for live location tracking. This app enables logistics fleets to manage their deliveries for the day. The app assigns deliveries to the driver, tracks live location, geofences arrival at a delivery location, and allows to add delivery notes and mark the delivery as complete. Fleet managers and teams can track the day's deliveries for all drivers in real-time on HyperTrack Views.

## Features

| Driver Checkin                 | Overview of Deliveries          | Delivery Details               |
| ------------------------------ | ------------------------------- | ------------------------------ |
| <img src="images/checkin.png"> | <img src="images/overview.png"> | <img src="images/details.png"> |

- Handle permission requests required for live location tracking
- Track day's deliveries as geofences on HyperTrack
- Automatically update arrival, exit and current delivery to make driver productive
- Create custom markers with delivery notes and for delivery completion
- Update delivery status to indicate time spent at the delivery destination using geofence webhooks

This app is using HyperTrack geofences, which come with powerful views on the **HyperTrack Dashboard** to:

- Track all devices associated with your HyperTrack account on a map in real-time
- Review all delivery details, including: delivery locations, time spent and route to each delivery
- Review custom markers with delivery notes and delivery completion to answer support tickets

<img src="images/dashboard.png">

## How to integrate the app with your backend

You can roll out the apps to drivers first and integrate backend later or do it the other way around.

### Associate app installs with drivers

Every time a driver installs the app and enters his `driver_id` a new `device_id` is created. Query `GET /devices` API to obtain a list of all app installs made by your drivers:
```bash
curl "https://v3.api.hypertrack.com/devices" \
     -u '{AccountId}:{SecretKey}'
```
* `AccountId` and `SecretKey` are API keys and can be obtained in the dashboard's [Setup page](https://dashboard.hypertrack.com/setup)

You will get a response similar to the following (check out all possible fields in our [API Reference](https://hypertrack.com/docs/references/#references-apis-devices-get-device-list):
```jsonc
[
    {
        "device_id": "DEADBEEF-DEAD-BEEF-DEAD-BEEFDEADBEEF",
        "registered_at": "2019-07-26T09:00:00Z",
        "metadata": {
            "driver_id": "driver@your_business.com"
        }
        ...
    }
]
```

Use the `driver_id` field to match a driver to a record in your system. Associate and store the `device_id` in your datastore. This identifier is important and it is used throughout HyperTrack APIs to identify devices. See [this guide](https://hypertrack.com/docs/guides/setup-and-manage-devices) for more about device management.

Run this query every time you expect drivers to install the Visits app. We recommend to make a query at the start of each working day before delivery assignment. This way your system will always be in sync with the app.

Apps can be reinstalled or installed on multiple phones, which can result in multiple devices with the same metadata but different `device_id`s. If you expect your drivers to only use one mobile phone with the app installed, update your user record with `device_id` having the latest `registered_at` value and delete the no longer needed `device_id` from HyperTrack. The deleted `device_id` won't be able to accidentally track and won't be billed for your account going forward, but the tracking data will be saved.

```bash
curl -X "DELETE" "https://v3.api.hypertrack.com/devices/{device_id}" \
     -u '{AccountId}:{SecretKey}'
```

### Assign deliveries to drivers

Every time you assign a delivery to a driver in your system create a geofence at customer's location.

```bash
curl -X "POST" "https://v3.api.hypertrack.com/geofences" \
     -u '{AccountId}:{SecretKey}' \
     -d $'{
    "device_ids": [
        "DEADBEEF-DEAD-BEEF-DEAD-BEEFDEADBEEF"
    ],
    "geofences": [
        {
            "geometry": {
                "type": "Point",
                "coordinates": [ -122.395223, 37.7947633]
            },
            "radius": 100,
            "metadata": {
                "Customer Notes": "Please deliver before 10 AM",
                "Message": "Use back door to enter."
            }
        }
    ]
}'
```

* `device_id` is the HyperTrack's id associated with your driver
* `geofences` you can assign multiple deliveries for the same driver at once by having multiple objects in this array
* `geometry`  is a geometry of a geofence with circular represented by "Point" value and polygon as "Polygon". Both types are represented in [GeoJSON format](https://hypertrack.com/docs/references#references-geojson) where coordinates have longitude as a 0'th element and latitude as 1'st
* `radius` is the radius for a circular geofence type. We suggest radius at least 100 meters to account for a driver meeting a customer outside of the building
* `metadata`  You can add any number of fields in the form `"string key": "string value"`  and the app will display them in the UI. You can also include non-string values, the app will ignore them, but you can use them later when querying the data or receiving events from webhooks.

Response:
```jsonc
[
    {
        "geofence_id": "deadbeef-dead-beef-dead-beefdeadbeef",
        "created_at": "2020-05-26T09:00:00.000Z",
        ...
    }
]
```

* `geofence_id` the ID of a newly created geofence. Associate it with your delivery and store it in your datastore
* `created_at` timestamp when geofence was created. Visits app uses this timestamp to order deliveries from oldest to newest

The geofence will be represented as a delivery on a driver's app. The delivery address will be reverse geocoded automatically by the app based on location. The street name and house number will be used as a name for the delivery.

### Track customer visits

HyperTrack uses webhooks for realtime events delivery. Follow the [setup guide](https://hypertrack.com/docs/guides/track-devices-with-api#stream-events-via-webhooks) to set them up.

Every time your driver enters or exits the customer's geofence a webhook event will be generated:
```jsonc
[
    {
        "type": "geofence",
        "device_id": "DEADBEEF-DEAD-BEEF-DEAD-BEEFDEADBEEF",
        "data": {
            "geofence_id": "deadbeef-dead-beef-dead-beefdeadbeef",
            "value": "exit",
            "arrival": {
                "location": {
                    "type": "Point",
                    "coordinates": [ 122.395235, 37.7947645]
                },
                "recorded_at": "2020-05-26T10:00:00.000Z"
            },
            "exit": {
                "location": {
                    "type": "Point",
                    "coordinates": [ 122.395265, 37.7947671]
                },
                "recorded_at": "2020-05-26T10:05:00.000Z"
            }
        },
        ...
    }
]
```

* `type` type of webhook data, here it's a `geofence` type webhook, but there are webhooks for many [other useful events](https://hypertrack.com/docs/references#references-webhooks)
* `device_id` of a device associated with driver that generated the event
* `geofence_id` the id of the geofence associated with delivery
* `value` can be either `entry` or `exit`. In case of `entry`, only `arrival` field will be present, and in case of `exit` both `arrival` and `exit` fields will be present
* `arrival` the driver's location where he entered the customer's geofence
* `exit` the driver's location where he exited the customer's geofence

### Track driver's performance

For every delivery, driver can send multiple delivery notes and mark every delivery as completed once. These actions will be sent as webhooks in real time:
```jsonc
[
    {
        "type:": "custom_marker",
        "device_id": "DEADBEEF-DEAD-BEEF-DEAD-BEEFDEADBEEF",
        "recorded_at": "2020-05-26T10:03:00.000Z",
        "location": {
            "type": "Point",
            "coordinates": [ 122.395265, 37.7947671]
        },
        "data": {
            "metadata": {
                "geofence_id" : "deadbeef-dead-beef-dead-beefdeadbeef",
                "delivery_note" : "Door code changed to 123"
            }
            "route_to":{
                "distance": 500,
                "duration": 120,
                "start_location": {
                    "geometry": {
                        "type": "Point"
                        "coordinates": [ 122.395265, 37.7947671],
                    },
                    "recorded_at": "2020-05-26T10:02:00.000Z"
                }
            }
        }
    },
    {
        "data": {
            "metadata": {
                "geofence_id" : "deadbeef-dead-beef-dead-beefdeadbeef",
                "completed" : true
            }
            ...
        }
        ...
    }
]
```

* `type` of webhook data. Visits app sends delivery notes and delivery completions as [`custom_marker` event](https://hypertrack.com/docs/references#references-webhooks-custom-marker-payload)
* `device_id` of a device that generated the event, corresponds to driver
* `recorded_at` a timestamp when driver made the action to either complete or send delivery notes
* `location` of the driver when he performed the action
* `metadata` will always have a `geofence_id` associated with the delivery the driver was interacting with and either `delivery_note` of type String the driver entered in a text field or `"completed": true` if driver tapped the "complete" button
* `route_to` contains a location of previous `custom_marker` with `distance` and `duration` between them. Use this data for driver's productivity analysis based on when and where driver completed the delivery

### Remove deliveries

Once delivery is completed or the next morning when new deliveries are dispatched, delete geofences associated with them using DELETE API:

```bash
curl -X "DELETE" "https://v3.api.hypertrack.com/geofences/{geofence_id}" \
     -u '{AccountId}:{SecretKey}'
```


## Documentation

For detailed documentation of the APIs, customizations and what all you can build using HyperTrack, please visit the official [docs](https://docs.hypertrack.com).

## Contribute

Feel free to clone, use, and contribute back via [pull requests](https://help.github.com/articles/about-pull-requests/). We'd love to see your pull requests - send them in! Please use the [issues tracker](https://github.com/hypertrack/logistics-ios/issues) to raise bug reports and feature requests.

We are excited to see what live location feature you build in your app using this project. Do ping us at help@hypertrack.com once you build one, and we would love to feature your app on our blog!

## Support

Join our [Slack community](https://join.slack.com/t/hypertracksupport/shared_invite/enQtNDA0MDYxMzY1MDMxLTdmNDQ1ZDA1MTQxOTU2NTgwZTNiMzUyZDk0OThlMmJkNmE0ZGI2NGY2ZGRhYjY0Yzc0NTJlZWY2ZmE5ZTA2NjI) for instant responses. You can also email us at help@hypertrack.com.


CHANGELOG.md
# Changelog

## [0.9.13] - 2021-06-17
### Changed
- Updated to use HyperTrack SDK [v5.2.4](https://github.com/hypertrack/sdk-android/blob/master/CHANGELOG.md#524---2021-06-17)

## [0.9.12] - 2021-06-15
### Changed
- Updated to use HyperTrack SDK [v5.2.3](https://github.com/hypertrack/sdk-android/blob/master/CHANGELOG.md#523---2021-06-15)

## [0.9.11] - 2021-06-15
- Bugfixes

## [0.9.10] - 2021-06-14
### Fixed
- Fixed geofence device_id error

## [0.9.9] - 2021-06-14
### Fixed
- Fixed crash on login

## [0.9.8] - 2021-06-11
### Changed
- Updated to use HyperTrack SDK [v5.2.2](https://github.com/hypertrack/sdk-android/blob/master/CHANGELOG.md#522---2021-06-11)

## [0.9.7] - 2021-06-10
### Fixed
- Fixed crash on login

## [0.9.6] - 2021-06-10
### Fixed
- Tracking service is automatically restarts on reboot now
### Changed
- HyperTrack SDK updated to v5.2.1

## [0.9.5] - 2021-06-08
### Changed
- HyperTrack SDK updated to v5.2.0

## [0.9.4] - 2021-06-04
### Changed
- Clock In geotags removed

## [0.9.3] - 2021-06-03
### Changed
- Background permission rationale changed.

## [0.9.2] - 2021-06-02 (UNRELEASED)
### Changed
- Background permission rationale style changed.

## [0.9.1] - 2021-06-01 (UNRELEASED)
### Fixed
- Driver ID input screen on older Androids UI fix
- Hubspot integration company name

## [0.9.0] - 2021-05-31 (UNRELEASED)
### Added
- Background location permission to use scheduled tracking hours
### Fixed
- Geofences related changes

## [0.8.2] - 2021-05-27
### Fixed
- Geofence list pagination

## [0.8.1] - 2021-05-27
### Fixed
- Geofence list pagination

## [0.8.0] - 2021-05-26
### Added
- Hubspot integration

## [0.7.2] - 2021-05-06
### Added
- New launcher and notification icons
### Fixed
- NPE crash on lateinit property not being initialized.

## [0.7.1] - 2021-04-23
### Fixed
- A crash when no last location is returned from the OS

## [0.7.0] - 2021-04-22
### Added
- Create trip from the app (Where are you going appearence)
- CLOCK_IN / CLOCK_OUT geotags were removed

## [0.6.11] - 2021-04-20
### Added
- Copy button for profile data

## [0.6.10] - 2021-04-19
### Fixed
- Crash on places screen fixed for OPPO devices

## [0.6.9] - 2021-04-16
### Added
- Route info in place visit details
- Copy Visit ID button at Visit details screen

### Changed
- Places list is now sorted by last visit

### Fixed
- Minor text fixes

## [0.6.8] - 2021-04-15
### Fixed
- Crash on visits list fixed

## [0.6.7] - 2021-04-13
### Fixed
- Crash in case if there no internet connection fixed

## [0.6.6] - 2021-04-13
### Fixed
- Visits list text changes
- Local visit UX changes and fixes

## [0.6.5] - 2021-04-09
### Fixed
- Mock locations were disallowed

## [0.6.4] - 2021-04-08
### Added
- Added Place visits timeline

### Changed
- Changed Select location UX

### Fixed
- Fixed missing street number in address when creating geofence

## [0.6.3] - 2021-04-07
### Added
- Added Place creation screen

## [0.6.2] - 2021-04-06
### Updated
- HyperTrack SDK was updated to v4.11.0

## [0.6.1] - 2021-03-31
- Manual Visit creation enabled by default
- Bugfix

## [0.6.0] - 2021-03-31
- Added Places tab
- Added Place details

## [0.5.3] - 2021-03-26
- Minor Sign Up changes

## [0.5.2] - 2021-03-18
- Fixed a bug that could lead to crash and null error snackbar on a map view.

## [0.5.1] - 2021-03-16
- Bug with map focusing on 0,0 coordinates when no history available fixed

## [0.5.0] - 2021-03-16
- Sign Up for HyperTrack without leaving the app.
- Interactive timeline to review your daily history.
- Minor UI improvements

## [0.4.0] - 2021-03-11
- We made the map view default screen
- Tracking now starts automatically for the very first app launch.
- Some labels were replaced with icons to preserve space and improve usability
- Geotags payload was made self-explanatory

## [0.3.1] - 2021-03-03
- Daily stats in the Summary tab
- Profile tab was added to explore the associated data
- You can attach multiple photos to each visit
- UI update to achieve better usability

## [0.2.13] - 2021-01-26
- Added whitelisting prompt for Xiaomi, OnePlus and alike.
- Visit completion events have their expected location attached.

## [0.2.12] - 2021-01-26
- Switched to efficient geofences API with markers included
- Replased Gson and Mockito with Moshi and MockK counterparts
- Visit click crash was fixed

## [0.2.11] - 2021-01-26
- Notification about device being deleted
- Non-blocking visits refresh
- Pull to refresh instead of button click
- Pick-Up button configurability via deeplink parameter.
- Local visit UI fixed.
- Bugfixes

## [0.2.10] - 2021-01-26
- HyperTrack SDK updated to v4.9.0
- Gson nullability issue fixed

## [0.2.9] - 2020-12-21
- Proof of Deliver photo can be attached to each visit
- Crashlytics integration
- Timeouts were incremented to 30 secs.

## [0.2.8] - 2020-12-15
- Auto check in for Trips and Geofences
- Updated visit state model to match Pick Up -> Check In -> Check Out / Cancel graph
- Daily history view added

## [0.2.7] - 2020-11-27
- Added login with HyperTrack Dashboard credentials
- Misc behavior changes

## [0.2.6-devpoc] - 2020-11-04
- Fixed issue with manual visits configuration been ignored

## [0.2.6] - 2020-10-30
- Added local visits and driver id configurability via deeplink

## [0.2.6-rc01] - 2020-10-27
- CheckIn/CheckOut is only available if configured in deeplinks
- Driver id can be passed as a deeplink parameter

## [0.2.5] - 2020-10-06
- Trip metadata entries, that have their keys starts with "ht_" prefix, aren't shown in customer notes.

## [0.2.4] - 2020-10-01
- Fixed a crash on network error.




app/src/main/java/com/hypertrack/android/models/Order.kt
package com.hypertrack.android.models

import com.google.android.gms.maps.model.LatLng
import com.hypertrack.android.api.TripDestination
import com.hypertrack.android.models.local.OrderStatus
import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.format.FormatStyle
import java.util.*

@Suppress("UNCHECKED_CAST")
@JsonClass(generateAdapter = true)
data class Order(
    @Json(name = "order_id") val id: String,
    val destination: TripDestination,
    @Json(name = "status") val _status: String,
    @Json(name = "scheduled_at") val scheduledAt: String?,
    val estimate: Estimate?,
    @Json(name = "metadata") val _metadata: Map<String, Any>?,
) {

    val status: OrderStatus
        get() = OrderStatus.fromString(_status)

    val metadata: Map<String, Any>
        get() = (_metadata ?: mapOf())


}

@JsonClass(generateAdapter = true)
data class Estimate(
    @Json(name = "arrive_at") val arriveAt: String?,
    @field:Json(name = "route") val route: Route?
)

@JsonClass(generateAdapter = true)
data class Route(
    @field:Json(name = "remaining_duration") val remainingDuration: Int?
)


/*
* {
  "account_id": "3d58f8b3-45fc-49be-828d-b7cc81afdd3b",
  "order_id": "1",
  "metadata": null,
  "scheduled_at": "2000-01-01T00:00:00.000Z",
  "device_id": "252CB4DA-03EE-36D9-9081-952F8FD73BA5",
  "destination": {
    "address": "1",
    "geometry": {
      "type": "Point",
      "coordinates": [
        -95.4604032,
        37.6430412
      ]
    },
    "radius": 30
  },
  "share_url": "https://trck.at/vROKAOq",
  "status": "ongoing",
  "started_at": "2021-03-24T13:54:50.600Z",
  "service_time": 0,
  "delayed": null,
  "eta_relevance_data": {
    "reason": "Destination too far",
    "time_at_irrelevance": "2021-03-24T13:55:00.749Z",
    "status": false
  }
}
*
*
*
*
*
* */